lib/SILOptimizer/Differentiation/PullbackCloner.cpp

```diff
@@ -989,26 +1000,136 @@ class PullbackCloner::Implementation final
     // differentiated.
     if (applyInfoLookup == nestedApplyInfo.end()) {
+      auto *normalBlock = tai->getNormalBB();
+      assert(normalBlock->getNumArguments() == 1 &&
+             "Expected try apply to have a single result");
       // Must not be active.
-      // TODO: Do we need to check token result for begin_apply?
-      SILValue result = fai.getResult();
-      assert(!result || !getActivityInfo().isActive(result, getConfig()));
+      assert(!getActivityInfo().isActive(normalBlock->getArgument(0), getConfig()));
       return;
     }
+
+    // try_apply pullback produces value only on non-error path, therefore
+    // its pullback might be not available. Therefore we wrap pullback into Optional
+    // in the linear map tuple. Build a diamond-shaped CFG fragment here, emitting
+    // switch_enum to unwrap an Optional<PullbackType>. If pullback is present we
+    // emit a pullback call and propagate adjoints of arguments to successor block.
+    // Otherwise, there is no pullback call, so we only propagate adjoints of would-be
+    // arguments. So, the code looks like as follows:
+    // bb1(%35 : $Float, %36 : @owned $(_: Optional<@callee_guaranteed (Float) -> Float>)):
+    //   %37 = destructure_tuple %36
+    //   switch_enum %37, case #Optional.some!enumelt: bb8, case #Optional.none!enumelt: bb9
+    // bb8(%39 : @owned $@callee_guaranteed (Float) -> Float)
+    //   %40 = apply %39(%0) : $@callee_guaranteed (Float) -> Float
+    //   accumulate adjoints (using %35)...
+    //   %54 = load [trivial] %42
+    //   br bb10(%54)
+    // bb9:
+    //   br bb10(%35)
+    // bb10(%60 : $Float):
+    //   ...
+    auto loc = tai->getLoc();
+    auto optPullback = getPullbackTupleElement(tai);
+    auto *pbBB = builder.getInsertionBB();
+    auto *normalPbBB = pbBB->split(builder.getInsertionPoint());
+    auto *errorPbBB = getPullback().createBasicBlockAfter(normalPbBB);
+    auto *afterTryApplyPbBB = getPullback().createBasicBlockAfter(errorPbBB);
+
+    // Note that we cannot simply assign map value as DenseMap entry references
+    // could be invalidated on insertion, so map[a] = map[b] may trigger
+    // use-after-free
+    auto currentBT = blockTemporaries[pbBB];
+    blockTemporaries.insert({afterTryApplyPbBB, currentBT});
+    blockTemporaries[pbBB].clear();
+
+    auto pullback =
+      builder.createSwitchOptional(loc, optPullback,
+                                   normalPbBB, errorPbBB,
+                                   optPullback->getOwnershipKind());
+
+    SmallVector<SILValue, 2> adjArgs;
+    FullApplySite fai(tai);
+    auto *originalBB = tai->getParent();
     auto &applyInfo = applyInfoLookup->getSecond();
+    for (unsigned i : applyInfo.config.parameterIndices->getIndices()) {
+      unsigned argIdx = fai.getNumIndirectSILResults() + fai.getNumIndirectSILErrorResults() + i;
+      auto origArg = fai.getArgument(argIdx);
+        auto paramInfo = fai.getSubstCalleeConv().getParamInfoForSILArg(argIdx);
```

Comment:
Nit: wrong indentation here and below. Maybe it's worth just running `git clang-format` on the changes.